modules/host/sector.go

```diff
@@ -0,0 +1,280 @@
+		// until all options have been exhausted.
+		potentialFolders := h.storageFolders
+		emptiestFolder, emptiestIndex := emptiestStorageFolder(potentialFolders)
+		for emptiestFolder != nil {
+			sectorPath := filepath.Join(h.persistDir, emptiestFolder.uidString(), string(sectorKey))
+			err := ioutil.WriteFile(sectorPath, sectorData, 0700)
+			if err != nil {
+				// Writing to the storage folder resulted in an error, try the
+				// next storage folder by removing the current folder from the
+				// set.
+				//
+				// TODO: Document the disk failure for user reporting purposes.
+
+				// Remove the failed folder from the list of folders that can
+				// be tried.
+				if emptiestIndex == len(potentialFolders)-1 {
+					potentialFolders = potentialFolders[:emptiestIndex-1]
+				} else {
+					potentialFolders = append(potentialFolders[0:emptiestIndex], potentialFolders[emptiestIndex+1:]...)
+				}
+				// Try the next folder.
+				emptiestFolder, emptiestIndex = emptiestStorageFolder(potentialFolders)
+				continue
+			}
+
+			// Update the database to reflect the new sector.
+			usage := sectorUsage{
+				Expiry:        []types.BlockHeight{expiryHeight},
+				StorageFolder: emptiestFolder.UID,
+			}
+			emptiestFolder.SizeRemaining -= sectorSize
+			usageBytes, err = json.Marshal(usage)
+			if err != nil {
+				return err
+			}
+			return bsu.Put(sectorKey, usageBytes)
+		}
+
+		// There is at least one disk that has room, but the write operation
+		// has failed.
+		return errDiskTrouble
+	})
+	if err != nil {
+		return err
+	}
+	return h.save()
+}
+
+// removeSector will remove a sector from the host at the given expiry height.
+// If the provided sector does not have an expiration at the given height, an
+// error will be thrown.
+func (h *Host) removeSector(sectorRoot crypto.Hash, expiryHeight types.BlockHeight) error {
+	// Open the database so that the sector usage information can be updated.
+	return h.db.Update(func(tx *bolt.Tx) error {
+		// Update the database to reflect the new sector.
+		bsu := tx.Bucket(bucketSectorUsage)
+		sectorKey := h.sectorID(sectorRoot[:])
+		sectorUsageBytes := bsu.Get(sectorKey)
+		var usage sectorUsage
+		err := json.Unmarshal(sectorUsageBytes, &usage)
+		if err != nil {
+			return err
+		}
+
+		// If there are multiple entries in the expiry, the sector cannot be
+		// deleted from disk because there are other obligations relying on the
+		// sector.
+		if len(usage.Expiry) > 1 {
+			// Find any entry in the usage that's at the expiry height and
+			// remove it.
+			var i int
+			for i := 0; i < len(usage.Expiry); i++ {
+				if usage.Expiry[i] == expiryHeight {
+					break
+				}
+			}
+			if i == len(usage.Expiry) {
+				return errors.New("removing a sector that doesn't seem to exist")
+			}
+			if i == len(usage.Expiry)-1 {
+				usage.Expiry = append(usage.Expiry[0:i])
+			} else {
+				usage.Expiry = append(usage.Expiry[0:i], usage.Expiry[i+1:]...)
```

Comment:
you can combine these branches; `append(usage.Expiry[:i], usage.Expiry[i+1:])` works for the `i == len(usage.Expiry-1)` case